src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,216 @@
+						return op;
+					}
+				}
+
+				l = propogateConstants(l);
+				r = propogateConstants(r);
+				Operation newop = new Operation(op.getOperator(), l, r);
+				return newop;
+			}
+
+			System.out.println("We should never get to this point!");
+			return expression;
+		}
+
+		private Expression evaluate(Expression expression) {
+			if (expression instanceof Constant) return expression;
+
+			assert expression instanceof Operation;
+			Operation operation = (Operation) expression;
+			Expression l = operation.getOperand(0);
+			Expression r = operation.getOperand(1);
+
+			if (l instanceof Operation) l = evaluate(l);
+			if (r instanceof Operation) r = evaluate(r);
+
+			Operation.Operator op = operation.getOperator();
+			int rint;
+			int lint;
+			switch (op) {
+				case EQ:
+					lint = ((IntConstant) l).getValue();
+					rint = ((IntConstant) r).getValue();
+					if (lint == rint) {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(0));
+					} else {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(1));
+					}
+				case GT:
+					lint = ((IntConstant) l).getValue();
+					rint = ((IntConstant) r).getValue();
+					if (lint > rint) {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(0));
+					} else {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(1));
+					}
+				case LT:
+					lint = ((IntConstant) l).getValue();
+					rint = ((IntConstant) r).getValue();
+					if (lint < rint) {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(0));
+					} else {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(1));
+					}
+				case AND:
+					Operation lop = (Operation) l;
+					Operation rop = (Operation) r;
+					assert lop.getOperator().equals(Operation.Operator.EQ);
+					assert rop.getOperator().equals(Operation.Operator.EQ);
+					if (lop.equals(rop)) {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(0));
+					} else {
+						return new Operation(Operation.Operator.EQ, new IntConstant(0), new IntConstant(1));
+					}
+				default:
+					System.out.println("Should not get here!");
+					return expression;
+			}
+		}
+
+		private boolean onlyContainsConstants(Operation operation) {
+			Expression l = operation.getOperand(0);
+			Expression r = operation.getOperand(1);
+
+			if (!(l instanceof Operation || r instanceof Operation)) {
+				return (l instanceof Constant && r instanceof Constant);
+			}
+
+			boolean ret = true;
+			if (l instanceof Operation) {
+				ret = ret && onlyContainsConstants((Operation) l);
+			}
+			if (r instanceof Operation) {
+				ret = ret && onlyContainsConstants((Operation) r);
+			}
+			return ret;
+		}
+
+		// For debugging purposes
+		private void pushAndPrint(Expression e) {
+			// System.out.println("Pushing expression to stack: " + e);
```

Comment:
For future reference, the logger has already been set up for you :muscle: Instead of doing a `println` and commenting it out later, just log on a debug level.